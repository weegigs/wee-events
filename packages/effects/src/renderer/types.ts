import * as Effect from "effect/Effect";
import { Entity, Payload } from "@weegigs/events-core";

export interface RenderFailure {
  readonly description: string;
}

export type Renderer<R, P extends Payload, A extends Payload> = Effect.Effect<(entity: Entity<P>) => Effect.Effect<A, RenderFailure>, never, R>;
export namespace Renderer {
  export type IO<P extends Payload, A extends Payload = Payload> = Renderer<unknown, P, A>;
}

export type AnyRenderer<R> = Renderer<R, Payload, Payload>;
export namespace AnyRenderer {
  export type IO = AnyRenderer<unknown>;

  export const from = <R, P extends Payload, A extends Payload>(source: Renderer<R, P, A>): AnyRenderer<R> => {
    return source as unknown as AnyRenderer<R>;
  };
}
