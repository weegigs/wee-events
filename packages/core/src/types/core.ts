import * as z from "zod";

export type Primitive = string | number | boolean;
export type Value = Primitive | Payload | List;
export type EmptyPayload = { [member: string]: never };
export type Payload = { [member: string]: Value } | EmptyPayload;
export type List = Primitive[] | Payload[] | List[];

export namespace Payload {
  const primitive = z.union([z.string(), z.number(), z.boolean()]);
  const list: z.Schema<List> = z.lazy(() => z.union([z.array(primitive), z.array(payload), z.array(list)]));
  const value: z.Schema<Value> = z.lazy(() => z.union([primitive, list, payload]));
  const payload: z.Schema<Payload> = z.lazy(() => z.record(value));

  export const schema = payload;
  export const parse = (value: unknown): Payload => schema.parse(value);
}

export namespace EmptyPayload {
  export const schema = z.record(z.never());
}

export type AggregateId<Type extends string = string> = {
  type: Type;
  key: string;
};

export namespace AggregateId {
  export const schema = <Type extends string>(type: Type): z.Schema<AggregateId<Type>> =>
    z.object({
      type: z.literal(type),
      key: z.string().min(1),
    }) as unknown as z.Schema<AggregateId<Type>>;

  export const encode = (aggregate: AggregateId): string => `${aggregate.type}.${aggregate.key}`;
  export const decoder =
    (type: string) =>
    (aggregate: string): AggregateId => {
      const [_type, ...key] = aggregate.split(".");
      return schema(type).parse({ type: _type, key: key.join(".") });
    };

  export const parser =
    <Type extends string>(type: Type) =>
    (value: unknown) =>
      schema(type).parse(value);

  export function create<const T extends string>(type: T, key: string): AggregateId<T> {
    return { type, key } as const;
  }
}

export interface DomainEvent<Type extends string = string, Data extends Payload = Payload> {
  type: Type;
  data: Data;
}

export namespace DomainEvent {
  export const schema = <S extends z.Schema>(type: string, data: S) =>
    z.object({
      type: z.literal(type),
      data,
    });
}

export interface RecordedEvent<E extends DomainEvent = DomainEvent> {
  /**
   * The aggregate that events belongs to.
   */
  aggregate: AggregateId;

  /**
   * The revision for the aggregate.
   */
  revision: string;

  /**
   * Unique identifier representing this event. ULID format.
   */
  id: string;

  /**
   * Type of this event.
   */
  type: E["type"];

  /**
   * Representing when this event was created in the database system.
   */
  timestamp: string;

  /**
   * Data of this event.
   */
  data: E["data"];

  /**
   * Representing the metadata associated with this event.
   */
  metadata: RecordedEvent.Metadata;
}

export namespace RecordedEvent {
  export type Metadata = {
    causationId?: string;
    correlationId?: string;
  };
}

export type Revision = string;

export namespace Revision {
  export type Initial = "00000000000000000000000000";
  export const Initial = "00000000000000000000000000";

  export const isInitial = (revision: Revision): boolean => revision === Initial;

  export const schema = z.string().min(26).max(26);
}
