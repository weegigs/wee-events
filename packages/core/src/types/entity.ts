import * as z from "zod";

import { AggregateId, Payload, Revision } from "./core";

export interface Entity<T extends Payload = Payload> {
  aggregate: AggregateId;
  type: string;
  revision: Revision;
  state: T;
}

export namespace Entity {
  export const schema = <Type extends string, Data extends z.ZodTypeAny>(type: Type, state: Data) =>
    z.object({
      aggregate: AggregateId.schema(type),
      type: z.string().min(1),
      revision: Revision.schema,
      state,
    });

  export function create<State extends Payload>(
    aggregate: AggregateId,
    type: string,
    revision: Revision,
    state: State
  ): Entity<State> {
    return { aggregate, type, revision, state } as const;
  }
}
