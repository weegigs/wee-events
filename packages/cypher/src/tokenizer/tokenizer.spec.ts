import { describe, it, expect } from "vitest";
import { Payload } from "@weegigs/events-core";

import { TokenizeFunction, Tokenizer } from "./types";

const tokenize: TokenizeFunction = async (path, value) => {
  return `${path}|${value}`;
};

describe("tokenizer", () => {
  it("should tokenize a payload with simple values", async () => {
    const payload: Payload = {
      string: "string",
      number: 42,
      boolean: false,
    };

    const t = Tokenizer.create(tokenize);

    const result = await t.tokenize(payload);
    expect(result).toMatchSnapshot();
  });

  it("should tokenize a payload with nested payloads", async () => {
    const payload: Payload = {
      nested: {
        string: "string",
        number: 42,
        boolean: false,
      },
    };

    const t = Tokenizer.create(tokenize);

    const result = await t.tokenize(payload);
    expect(result).toMatchSnapshot();
  });

  it("should tokenize a payload with nested lists", async () => {
    const payload: Payload = {
      nested: [
        {
          string: "string",
          number: 42,
          boolean: false,
        },
        {
          string: "another",
          number: 96,
          boolean: true,
        },
      ],
    };

    const t = Tokenizer.create(tokenize);

    const result = await t.tokenize(payload);
    expect(result).toMatchSnapshot();
  });
});
