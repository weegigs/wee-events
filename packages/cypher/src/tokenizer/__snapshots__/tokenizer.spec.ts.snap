// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tokenizer > should tokenize a payload with nested lists 1`] = `
{
  "nested": [
    {
      "boolean": true,
      "number": 833454428,
      "string": "nested.0.string|string",
    },
    {
      "boolean": true,
      "number": -963496772,
      "string": "nested.1.string|another",
    },
  ],
}
`;

exports[`tokenizer > should tokenize a payload with nested payloads 1`] = `
{
  "nested": {
    "boolean": true,
    "number": 1681225690,
    "string": "nested.string|string",
  },
}
`;

exports[`tokenizer > should tokenize a payload with simple values 1`] = `
{
  "boolean": false,
  "number": 1649953809,
  "string": "string|string",
}
`;
